osf/metrics/counted_usage.py

```diff
@@ -0,0 +1,173 @@
+
+
+@receiver(pre_save, sender=CountedUsageV6)
+def autofill_fields(sender, instance, **kwargs):
+    pageview = getattr(instance, 'pageview_info', None)
+    if pageview:
+        fill_pageview_info(instance)
+    item_guid = getattr(instance, 'item_guid', None)
+    if item_guid:
+        guid_instance = Guid.load(item_guid)
+        if guid_instance and guid_instance.referent:
+            fill_osfguid_info(instance, guid_instance.referent)
+    fill_document_id(instance)
+
+
+def fill_pageview_info(counted_usage):
+    pageview = counted_usage.pageview_info
+    pageview.hour_of_day = counted_usage.timestamp.hour
+    pageview.page_path = urlsplit(pageview.page_url).path.rstrip('/')
+    referer = getattr(pageview, 'referer_url', None)
+    if referer:
+        pageview.referer_domain = urlsplit(referer).netloc
+
+
+def fill_osfguid_info(counted_usage, guid_referent):
+    counted_usage.item_public = get_ispublic(guid_referent)
+    counted_usage.item_type = ContentType.objects.get_for_model(
+        guid_referent.__class__,
+        for_concrete_model=False,
+    ).model
+    counted_usage.surrounding_guids = get_surrounding_guids(guid_referent)
+    if not counted_usage.provider_id:
+        counted_usage.provider_id = get_provider_id(guid_referent)
+
+
+def fill_document_id(counted_usage):
+    if counted_usage.pageview_info is not None:
+        target_identifier = counted_usage.pageview_info.page_url
+    else:
+        target_identifier = counted_usage.item_guid
+
+    # divide the day into 30-second windows, filter out
+    # "double clicks" by removing duplicates within a window
+    # (latest event wins, per COUNTER)
+    day_start = datetime(
+        counted_usage.timestamp.year,
+        counted_usage.timestamp.month,
+        counted_usage.timestamp.day,
+        tzinfo=pytz.utc,
+    )
+    time_in_seconds = (counted_usage.timestamp - day_start).total_seconds()
+    time_window = int(time_in_seconds / 30)
+
+    # Set the document id to a hash of "unique together"
+    # values to get "ON CONFLICT UPDATE" behavior -- if
+    # the document already exists, it will be updated rather
+    # than duplicated.  Cannot detect/avoid conflicts this way,
+    # but that's ok because we want "doubleclick filtering"
+    counted_usage.meta.id = stable_key(
+        counted_usage.platform_iri,
+        target_identifier,
+        counted_usage.session_id,
+        counted_usage.timestamp.date(),
+        time_window,
+    )
+
+
+def get_ispublic(guid_referent):
+    # if it quacks like BaseFileNode, look at .target instead
+    maybe_public = getattr(guid_referent, 'target', None) or guid_referent
+    return bool(
+        getattr(maybe_public, 'verified_publishable', None)  # quacks like Preprint
+        or getattr(maybe_public, 'is_public', None)          # quacks like AbstractNode
+    )
+
+
+def get_provider_id(guid_referent):
+    provider = getattr(guid_referent, 'provider', None)
+    if isinstance(provider, str):
```

Comment:
especial thanks because i'm pretty sure the last case `getattr(guid_referent, 'provider_id', None)` won't ever happen (not sure what my thought was) -- files have `provider = 'foo'`, preprints and registrations have `provider=FooProvider()`, and i added a fall-thru `return 'osf'` for the rest